scico/numpy/blockarray.py

```diff
@@ -0,0 +1,352 @@
+
+        return super().__init__(arrays)
+
+    def _full_ravel(self) -> DeviceArray:
+        """Return a copy of ``self._data`` as a contiguous, flattened `DeviceArray`.
+
+        Note that a copy, rather than a view, of the underlying array is
+        returned. This is consistent with :func:`jax.numpy.ravel`.
+
+        Returns:
+            Copy of underlying flattened array.
+
+        """
+        return jnp.concatenate(tuple(x_i.ravel() for x_i in self))
+
+    @property
+    def dtype(self):
+        """Allow snp.zeros(x.shape, x.dtype) to work."""
+        return self[0].dtype
+
+    def __getitem__(self, key):
+        """Make, e.g., x[:2] return a BlockArray, not a list."""
+        result = super().__getitem__(key)
+        if not isinstance(result, jnp.ndarray):
+            return BlockArray(result)
+        return result
+
+    """ backwards compatibility methods, could be removed """
+
+    @staticmethod
+    def array(iterable):
+        """Construct a :class:`.BlockArray` from a list or tuple of existing array-like."""
+        return BlockArray(iterable)
+
+
+# Register BlockArray as a jax pytree, without this, jax autograd won't work.
+# taken from what is done with tuples in jax._src.tree_util
+jax.tree_util.register_pytree_node(
+    BlockArray,
+    lambda xs: (xs, None),  # to iter
+    lambda _, xs: BlockArray(xs),  # from iter
+)
+
+""" Wrap unary ops like -x. """
+
+
+def _unary_op_wrapper(op):
+    op = getattr(DeviceArray, op_name)
+
+    @wraps(op)
+    def op_ba(self):
+        return BlockArray(op(x) for x in self)
+
+    return op_ba
+
+
+for op_name in unary_ops:
+    setattr(BlockArray, op_name, _unary_op_wrapper(op_name))
+
+""" Wrap binary ops like x+y. """
+
+
+def _binary_op_wrapper(op_name):
+    op = getattr(DeviceArray, op_name)
+
+    @wraps(op)
+    def op_ba(self, other):
+        if isinstance(other, BlockArray):
+            return BlockArray(op(x, y) for x, y in zip(self, other))
+
+        result = BlockArray(op(x, other) for x in self)
+        if NotImplemented in result:
+            return NotImplemented
+        return result
+
+    return op_ba
+
+
+for op_name in binary_ops:
+    setattr(BlockArray, op_name, _binary_op_wrapper(op_name))
+
+
+""" Wrap DeviceArray properties. """
+
+
+def _da_prop_wrapper(prop_name):
+    prop = getattr(DeviceArray, prop_name)
+
+    @property
+    @wraps(prop)
+    def prop_ba(self):
+        result = tuple(getattr(x, prop_name) for x in self)
+        if isinstance(result[0], jnp.ndarray):
+            return BlockArray(result)
+        return result
+
+    return prop_ba
+
+
+skip_props = ("at",)
+da_props = [
+    k
+    for k, v in dict(inspect.getmembers(DeviceArray)).items()
+    if isinstance(v, property) and k[0] != "_" and k not in dir(BlockArray) and k not in skip_props
+]
+
+for prop_name in da_props:
+    setattr(BlockArray, prop_name, _da_prop_wrapper(prop_name))
+
+""" Wrap DeviceArray methods. """
```

Comment:
These section dividers don't work so well because of the single space above and double space below. I suspect that `black` is insisting on the double space below? Can a triple space be added above, or would that also be removed by `black`?

Reply:
Changed to normal line comments, which seem to work better.